tests/transforms/test_compile.py

```diff
@@ -0,0 +1,406 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import pytest
+from pennylane import numpy as np
+
+import pennylane as qml
+from pennylane.wires import Wires
+
+from pennylane.transforms.optimization import cancel_inverses, commute_controlled, merge_rotations
+from pennylane import compile
+
+from test_optimization.utils import compare_operation_lists
+
+
+def build_qfunc(wires):
+    def qfunc(x, y, z):
+        qml.Hadamard(wires=wires[0])
+        qml.RZ(z, wires=wires[2])
+        qml.CNOT(wires=[wires[2], wires[1]])
+        qml.CNOT(wires=[wires[1], wires[0]])
+        qml.RX(x, wires=wires[0])
+        qml.CNOT(wires=[wires[1], wires[0]])
+        qml.RZ(-z, wires=wires[2])
+        qml.RX(y, wires=wires[0])
+        qml.PauliY(wires=wires[2])
+        qml.CY(wires=[wires[1], wires[2]])
+        return qml.expval(qml.PauliZ(wires=wires[0]))
+
+    return qfunc
+
+
+class TestCompile:
+    """Test that compilation pipelines work as expected."""
+
+    @pytest.mark.parametrize(("wires"), [["a", "b", "c"], [0, 1, 2], [3, 1, 2], [0, "a", 4]])
+    def test_empty_pipeline(self, wires):
+        """Test that an empty pipeline returns the original function."""
+
+        qfunc = build_qfunc(wires)
+        dev = qml.device("default.qubit", wires=wires)
+
+        qnode = qml.QNode(qfunc, dev)
+
+        transformed_qfunc = compile(pipeline=[])(qfunc)
+        transformed_qnode = qml.QNode(transformed_qfunc, dev)
+
+        original_result = qnode(0.3, 0.4, 0.5)
+        transformed_result = transformed_qnode(0.3, 0.4, 0.5)
+        assert np.allclose(original_result, transformed_result)
+
+        names_expected = [op.name for op in qnode.qtape.operations]
+        wires_expected = [op.wires for op in qnode.qtape.operations]
+
+        compare_operation_lists(transformed_qnode.qtape.operations, names_expected, wires_expected)
+
+    @pytest.mark.parametrize(("wires"), [["a", "b", "c"], [0, 1, 2], [3, 1, 2], [0, "a", 4]])
+    def test_default_pipeline(self, wires):
+        """Test that the default pipeline returns the correct results."""
+
+        qfunc = build_qfunc(wires)
+        dev = qml.device("default.qubit", wires=Wires(wires))
+
+        qnode = qml.QNode(qfunc, dev)
+
+        transformed_qfunc = compile()(qfunc)
+        transformed_qnode = qml.QNode(transformed_qfunc, dev)
+
+        original_result = qnode(0.3, 0.4, 0.5)
+        transformed_result = transformed_qnode(0.3, 0.4, 0.5)
+        assert np.allclose(original_result, transformed_result)
+
+        names_expected = ["Hadamard", "CNOT", "RX", "CY", "PauliY"]
+        wires_expected = [
+            Wires(wires[0]),
+            Wires([wires[2], wires[1]]),
+            Wires(wires[0]),
+            Wires([wires[1], wires[2]]),
+            Wires(wires[2]),
+        ]
+
+        compare_operation_lists(transformed_qnode.qtape.operations, names_expected, wires_expected)
```

Comment:
This is a good test to have, but it is very much an integration test - that is, it is testing that the compile function _and_ all pipeline functions work as intended. It might be good to write a few unit tests in addition, e.g.:

- Have a test that uses pytest-mock to 'spy' on the pipeline functions, and ensure that they have been called the correct number of times with the correct number of arguments, but do not compare the function outputs.

Reply:
Okay this makes a lot more sense, since the transforms themselves have already been tested extensively. I'll add a few new tests like this!

Reply:
Really useful to see these thoughts on how testing is approached at this level of functionality!